Added `.equals()` and `.hashCode()` to all AbstractArray subclasses

nedtwigg opened the pull request:
Straightforward equality for all the `AbstractStruct` and `AbstractMatFile` subclasses.  The only wrinkle is that one of the MCOS examples actually creates a cyclic object graph, which causes the equality check to StackOverflow.  A cyclic object graph inside a MAT-file seems like enough of a corner case that the feature is valuable on its own, not clear how much effort it's worth expending to dig into such a tiny case.

ennerf commented:
Thank you for the PR. I'll be travelling until next Thursday, but I'll try to take a look at it as soon  as I can.

ennerf commented on src/main/java/us/hebi/matlab/mat/format/NumberStore.java:
```diff
@@ -48,4 +52,73 @@
+                return aCast.buffer.equals(bCast.buffer);
+            }
+        }
+        if (logical) {
```
May result in different behavior for non-`UniversalNumberStore`, i.e., logical types may result in equal results even if the bytes in the buffer are different.

nedtwigg replied:
This is intentional.  I have two MAT-files that load identically into MATLAB, but have different bytes.  They are both `double` in MATLAB, but one is backed by `uint8`, and the other by `double` in the data of the MAT-File.  Don't know how they got generated this way, but they're there :)  I'd love to contibute them as testcases, but they have customer data, and I don't know how to make them have different backing datatypes.

I considered removing the "fastcase" (premature optimization?) but it's so easy I left it in.

ennerf replied:
I understood the intention, but I apparently missed the `if (aCast.type == bCast.type)` line. Your implementation is correct and makes sense.

ennerf commented on src/main/java/us/hebi/matlab/mat/types/AbstractStruct.java:
```diff
@@ -111,4 +111,37 @@ public void close() throws IOException {
+        }
+    }
+
+    protected abstract boolean subSubEquals(Object otherGuaranteedSameClass);
```
I can see a reason for the first `sub-` level, but nesting them seems a bit much. I'd prefer children to just override the `subEqualsGuaranteedSameClass` and call `super` if necessary.

If you feel strongly about it and really want to enforce explicit implementation, the `AbstractStruct::subEqualsGuaranteedSameClass` should be `final`.

nedtwigg replied:
Roger.  Subclassing and equality are tricky, so I like the `final` approach.

nedtwigg replied:
Nvm, I think you are right that calling `super.` is better, I misunderstood the class hierarchy.

ennerf commented on src/test/java/us/hebi/matlab/mat/tests/mat5/SimulinkTest.java:
```diff
@@ -41,7 +41,11 @@
+        // the object graph is not hierarchical or even a DAG.
+        // it is a cyclic graph which causes StackOverflow on equality check
+        // rare enough that it's not clear how much investigation is warranted...
+        boolean equalityCheck = false;
```
I'm ok punting on equality cycles.

ennerf commented on src/main/java/us/hebi/matlab/mat/format/NumberStore.java:
```diff
@@ -20,11 +20,15 @@
 import java.io.Closeable;
 import java.io.IOException;
 
+import javax.annotation.Nullable;
```
Not available on maven build

nedtwigg commented on src/main/java/us/hebi/matlab/mat/format/Mat5File.java:
```diff
@@ -302,4 +303,18 @@ private static void checkMat5Identifier(String description) {
 
+    @Override
+    protected int subHashCode() {
+        return Objects.hash(description, subsysOffset, byteOrder, version, reduced);
```
Roger.  Re: Java 6 compliance, is this a requirement for you, or is it a requirement to make it easier for others to adopt?  I think Java 8 penetration is pretty good now (it was [75% in Feb 2018](https://www.baeldung.com/java-in-2017)).  Gradle 5 now requires Java 8.

Happy to stick to SE6 if you want :)

ennerf replied:
We have an API that uses it and should be able to run in MATLAB 2012b (bundles a Java SE 6 runtime). Unless there is a significant upside to it, I'd prefer to not change the requirement to 2013b (bundles Java SE 7 runtime) quite yet.

nedtwigg commented:
Thanks for the feedback!  I addressed all the points, ready for the next round :)  `mvn test` passes on my box, can't figure out how to see the CI error.

ennerf commented:
Accepted. Thank you for the PR!

The previous errors were related to `@Nullable`, but the latest one is a build issue on macOS only. I believe it's related to a new worker machine that was added today and has a different setup than the others. I'll look into it. If you happen to be an expert with Travis I'd be happy to accept a PR for a working script as well :)

ennerf commented:
Thanks! Does Travis not support Maven?

I just released the latest changes as version `0.3`. I think I'll do the `1.0` after addressing #20 and #21, which will hopefully be next week. This may break backwards compatibility on `NamedArray` and `global`, but it's probably already ok to link to `Mat-File IO`.
